Adding gene sets documentation to cellxgene docs

signechambers1 opened the pull request:
Adding gene sets documentation and sample CSV for pip installed cellxgene to docs.

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+## Overview
+
+Users may view gene sets along with their dataset and access gene sets created in pip installed cellxgene by following the instructions below.
```
I suggest defining a `geneset` before using it.   Ie, "In cellxgene, a geneset is ...`

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+Users may view gene sets along with their dataset and access gene sets created in pip installed cellxgene by following the instructions below.
+
+The order of the gene symbols in each gene set is maintained in the database. 
```
what is `the database`?  Put another way, what is it that this sentence is trying to communicate?

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED" "MAY", and "OPTIONAL" in this document are to be interpreted as described in BCP 14, [RFC2119](https://www.rfc-editor.org/rfc/rfc2119.txt), and [RFC8174](https://www.rfc-editor.org/rfc/rfc8174.txt) when, and only when, they appear in all capitals, as shown here.
+
+## cellxgene CLI commands
```
This section needs to clarify interactions with multiple users and the session cookie.

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+## Rows
+
+Each subsequent row describes a gene in a gene set.
```
I would add a statement about uniqueness constraints here, eg, "Each row contains a unique [geneset, gene] tuple."

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+  
+The cellxgene gene set data format is a [*Tidy* CSV](./gene_sets_example.csv) (comma-separated values) file using ASCII encoding. Multiple gene sets MAY be included in the file similar to the [Gene Matrix Transposed](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29) or [Gene Matrix](https://software.broadins ...
+
+Example:
```
I find the extraneous (non-normative) columns confusing.  I would keep your initial example minimal (eg, delete `provenance1...`).  If you feel that the prose in the Mandatory Header section needs further examples to highlight the allowance for "additional columns", add it there.

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+## Mandatory Header
+
+The first row MUST contain header columns using reserved names in the following order:
```
What is a `header column`?  I think you just mean `column`?

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+Each subsequent row describes a gene in a gene set.
+
+---
```
Formatting gets a bit wonky below here.  It is organized like the subsequent sections are sub-sections of `Rows`, but in fact, they are the definition of `Columns`.   Suggest adding a header, eg, `Column Specifications`

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+### `gene_set_name`
+
+The `gene_set_name` column MUST contain a value. The values for `gene_set_name` MUST NOT contain illegal ASCII characters or sequences. If the following cases are detected, validation MUST display an error message and fail the upload:
```
What does "fail the upload" and "validation" imply?   AFAIK, there is no "upload" in cellxgene, and it makes it sound like validation is a separate procedural step (if so, need to introduce it).

Instead of the above, if this is a data format specification, just state what is legal/illegal, and leave the UI/workflow out of it.  If it is a user guide, you will need to explain a bit more.

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+## cellxgene gene set CSV data format
+  
+The cellxgene gene set data format is a [*Tidy* CSV](./gene_sets_example.csv) (comma-separated values) file using ASCII encoding. Multiple gene sets MAY be included in the file similar to the [Gene Matrix Transposed](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29) or [Gene Matrix](https://software.broadins ...
```
Is the format really ASCII?  Or is it UTF-8/Unicode?  I was pretty sure it was unicode, but I may be wrong....

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+### `gene_symbol`
+
+The `gene_symbol` column MUST contain a value that is unique for the `gene_set_name`. Upload will fail and display an error message for a duplicate `gene_symbol`. This is illustrated by **~~CCKAR~~** in the example:
```
IIRC, the current application does not generate an error for duplicate gene_set/gene tuples, but instead just drops duplicates.   (caveat: comment on behavior is from memory)

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+### `gene_description`
+
+The `gene_description` column MAY contain a value and will be surfaced when a user hovers on `gene_symbol` in cellxgene.
```
this doc can't decide if it is a user guide (UI, etc) or a data format spec.  So it is struggling on both cases :-)

If you need both, I suggest two docs, or at least two completely separate sub-docs.

For example, from a data format viewpoint, this section should just say "gene description is optional", and the UI is non-normative

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,103 @@
+
+The `gene_description` column MAY contain a value and will be surfaced when a user hovers on `gene_symbol` in cellxgene.
+
+## Creating and editting gene sets using cellxgene UX
```
this also should not be in the file spec.  Some implementations will let the user edit, some will not...

Also, if this is a user guide, the CLI params have a lot of impact on where data is written (so "created in the user's home directory" may be incorrect).

bkmartinjr reviewed (changes requested):
Suggest an editing/refinement pass, after making a decision if this is a format specification or a user guide.   If the former, there are many non-normative sections that should be removed.  If the latter, the normative prose could be removed for clarity, and more examples provided.

signechambers1 commented:
@bkmartinjr I lost my mind and forgot to commit changes to this PR last week. I've updated the text to read more like a user guide. Would you mind taking another look?

Please pay extra attention to the multi-user language, I added a little bit under the CLI Command section but tried to keep it simple. If you think there is extra info that would be helpful let me know!

@MaximilianLombardo I'm counting your review as optional. If you want to be required for me to merge let me know!

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,110 @@
+## Overview
+
+A gene set is a named list of user-ordered genes. A gene set can describe marker genes for a cell type, state, or pathway.
```
I'm not sure if `user-ordered genes` is the clearest concept...it may be clearer to say "ordered list of genes".  

Suggest something like:   

> A gene set is a user-created, ordered list of genes identifiers.  Typical uses for a gene set include marker genes for a cell type, state, or pathway.  
> 
> A gene set must have a unique name.  A given gene may be included in multiple gene sets. A gene set may optionally include a description for the entire gene set, and descriptions for each gene included the gene set.

bkmartinjr commented on docs/gene_sets.md:
```diff
@@ -0,0 +1,110 @@
+
+### `gene_symbol`
+
+`gene_symbol` MUST contain a unique gene symbols for the given `gene_set_name` and exist as a VAR in the underlying anndata file.
```
I would add:

> A given gene_symbol may only be added once to a gene set.

(or something to that effect - ie, you can't have dups)

bkmartinjr commented on docs/gene_sets_example.csv:
```diff
@@ -0,0 +1,6 @@
+club.cell,Genes that when expressed together uniquely define club cells,SCGB3A2,protein coding gene associated with asthma,Pubmed ID ABC456,Primary Pubmed ID,Pubmed ID 2928374,Secondary Pubmed ID,
+club.cell,Genes that when expressed together uniquely define club cells,CYP2F2,protein coding gene associated with metabolic processes,Pubmed ID DCF678,Primary Pubmed ID,Pubmed ID XYZ123,Secondary Pubmed ID,
+macrophage,Genes that when expressed together uniquely define macrophages,CD68,Critical to macrophages,,,,,
+macrophage,Genes that when expressed together uniquely define macrophages,CD163,Also critical to macrophages,,,,,
```
missing newline at end of file.

bkmartinjr reviewed (approved):
Three small items.  But otherwise it looks great.  Feel free to merge after you look at the three items.
